apps/aestratum/src/aestratum.erl

```diff
@@ -0,0 +1,447 @@
+
+-include("aestratum.hrl").
+-include("aestratum_log.hrl").
+-include_lib("aecontract/include/aecontract.hrl").
+-include_lib("aecore/include/blocks.hrl").
+-include_lib("stdlib/include/ms_transform.hrl").
+
+-import(aestratum_fn, [ok_err/2, ok_val_err/1, ok_val_err/2]).
+
+-define(PAYOUT_CHECK_INTERVAL,            timer:minutes(2)).
+-define(CHAIN_PAYMENT_TX_CHECK_INTERVAL,  timer:seconds(10)).
+-define(CANDIDATE_CACHE_CLEANUP_INTERVAL, timer:minutes(15)).
+
+-define(PAYMENT_CONTRACT_INIT_GAS, 250).           % measured
+-define(PAYMENT_CONTRACT_GAS_PER_TRANSFER, 22000). % measured
+
+%%%===================================================================
+%%% STATE
+%%%===================================================================
+
+-record(reward_key_block,
+        {share_key :: aestratum_db:sort_key(),
+         height    :: non_neg_integer(),
+         target    :: non_neg_integer(),
+         tokens    :: aestratum_db:amount(),
+         hash      :: binary()}). %% reward block hash
+
+-record(aestratum_state,
+        {height,
+         balance,
+         tx_push_pid,
+         pending_tx_id,
+         pending_tx_hash}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+submit_share(<<"ak_", _/binary>> = Miner, MinerTarget, <<CandidateBlockHash/binary>>) ->
+    true = is_integer(MinerTarget) andalso MinerTarget >= 0,
+    gen_server:cast(?MODULE, {submit_share, Miner, MinerTarget, CandidateBlockHash}).
+
+submit_solution(CandidateBlockHash, Nonce, Pow) ->
+    gen_server:cast(?MODULE, {submit_solution, CandidateBlockHash, Nonce, Pow}).
+
+status() ->
+    ConnPids = aestratum_user_register:conn_pids(),
+    [aestratum_handler:status(ConnPid) || ConnPid <- ConnPids].
+
+status(Account) when is_binary(Account) ->
+    case aestratum_user_register:find(Account) of
+        {ok, #{conn_pid := ConnPid}} -> aestratum_handler:status(ConnPid);
+        {error, _Rsn} = Err          -> Err
+    end.
+
+db_keys(Height) ->
+    ?TXN(aestratum_db:keys(Height)).
+
+tab_keys() ->
+    ?TXN(lists:foldl(fun (T, Acc) -> Acc#{T => mnesia:all_keys(T)} end, #{}, ?TABS)).
+
+tab_keys(Tab) ->
+    ?TXN(mnesia:all_keys(Tab)).
+
+sent_payments() ->
+    ?TXN(aestratum_db:select_payments(true)).
+
+pending_payments() ->
+    ?TXN(aestratum_db:select_payments(false)).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+init([]) ->
+    PoolRelShares = aestratum_reward:relative_shares(?POOL_PERCENT_SHARES, ?POOL_PERCENT_SUM),
+    aestratum_env:set(#{pool_relative_shares => PoolRelShares}),
+    ?TXN(aestratum_db:is_empty(?ROUNDS_TAB) andalso aestratum_db:store_round()),
+    aec_events:subscribe(stratum_new_candidate),
```

Comment:
How can this init know that this subscription does not sometimes cause runtime error, considering that aestratum app does not depend on any other non-Stratum Aeternity app or gproc?

Reply:
I've added dependency on aecore.